### DISCUSSION TITLE
Reporting order during filtering

### DISCUSSION THREAD

**Discussion by TonyGravagno** [RealRaven2000/FiltaQuilla#294] [General]
I'm observing patterns in the order in which filters are processed that might be of interest to folks here.

### Processing rules (search terms) for a true/false

1. The Browser Console in Tb "seems" to show filters are processed synchronously. When processing multiple mail items with multiple filters, quickFilters console.log entries show search term 1 for a message, then term 2, then term 3. One millisecond later it reports the same correct sequence for the next mail item being processed. While some terms are processed in the same millisecond, they are always in correct order in the log.
2. In the Browser Console, there is a gear setting for Group Similar Messages. Even if this is disabled, logging "somehow" gets grouped. I have a log here that shows individual FiltaQuilla logs as described in 1 above. Then there is a gap where I am using Tb terms, then there is another FQ log entry ... but it shows four logs grouped as one. This could only happen if four mail items were being processed by the same search term at the same time. That means processing is Asynchronous, where multiple email items are being processed at the same time, and can be processed by the same filter and the same term at the same time, resulting in a grouping of identical (timestamped) log entries.

So are messages being processed synchronously or asynchronously? It's not going to change while processing so it seems like it's async. Earlier in the log I believe entries look synchronous (and they actually are at this moment) as Tb begins processing a list of message items. It queues the first message, then the second, etc. The first one zips through processing while the second is being queued, so it's processing in this area is truly synchronous. As more items get added to the queue, the same rules (and FQ logging) may or may not be executed simultaneously for different mail items, giving the impression that the same rules might be run on a single mail item or that other rules are being processed out of order.

The only way to know which message is being processed, and which filter is being processed on that message, is if FQ includes this data. I've created a separate discussion here on that topic.

Unless we can find an event for Tb processing a single filter or term within a filter, Tb does not report its filter processing to the log. (I recently disabled many debug/filter prefs and might have simply turned this off.) This leads to the idea that if you really want to see detailed logging for filters, you need to change standard Tb terms to filtaQuilla terms, and FQ needs to provide detail that prevents logs from getting grouped.

### Processing actions when terms match

3. The Filter Log does not show term matches. It shows actions. The log entries show the time (without milliseconds), a filter name, message info, and each **action** taken by a filter for a message.

- It shows a first/only in a single filter being applied to multiple messages in the same second. (Without a Stop action)
- Then, in the same message order, it shows the first action of another filter being applied to all matching messages.
- Then it shows the second action of this second filter being applied to all messages, in the same order.

To clarify: Filter1 Action1 gets applied to messages 1,2,3,4. Then Filter2 Action1 gets applied to messages 1,2,3,4. Then Filter2 Action2 gets applied to messages 1,2,3,4.

### Why is this worthy of a note?

It confirms that a single message gets run through all terms before another message is processed. And yet for actions, it processes a single action on multiple messages, then the next action on all messages, etc.

If you create JavaScript actions or Run Program actions in FQ that are dependent on one another, it's not going to behave as you expect unless you are aware of these details. I asked in #274 about preserving state between JavaScript actions. Based on the information here, it seems after a script is executed, another message is processed ... so state between scripts isn't going to apply to a single message. If we can pass state from one action to another (Tb storage?) we must create a table of message IDs to ensure we know which state applies to each message.

Observation 3 might be old info to some it was certainly new to me. This is another tool in the toolbox : track your rules in the console, be aware of async processing, don't expect logs for Tb rules, and then check results in the Filter Log ... not the console.

How accurate is all of this? I dunno, but until proven otherwise this is what I have now learned to expect from this software.   
Please comment with confirmation or different observations.

**Comment by RealRaven2000** [comment]
> So are messages being processed synchronously or asynchronously? It's not going to change while processing so it seems like it's async. Earlier in the log I believe entries look synchronous (and they actually are at this moment) as Tb begins processing a list of message items. It queues the first message, then the second, etc. The first one zips through processing while the second is being queued, so it's processing in this area is truly synchronous. As more items get added to the queue, the same rules (and FQ logging) may or may not be executed simultaneously for different mail items, giving the impression that the same rules might be run on a single mail item or that other rules are being processed out of order.

Message filters are processed synchronously. But a synchronous action may call (start) and asynchronous function - it just cannot await the end of said function. Lengthy operations ( and a lot of future functionality) will be converted to async in the future so that the process queue gets relieved and the Computer may remain responsive. If a file operation such as moving a mail or extracting attachments is executed (started) asynch from a synchronous process such as a filter action, we cant assume that the operation is complete by the time the filter action reports that it is complete:

In this example, (only assuming these are async operations) it is possible.that Thunderbird will execute the physical move / attachment store / removal etc. at a later stage, which may lead to problems if there are other filters following that may need the message content in a certain state.

> In the Browser Console, there is a gear setting for Group Similar Messages. Even if this is disabled, logging "somehow" gets grouped. I have a log here that shows individual FiltaQuilla logs as described in 1 above. Then there is a gap where I am using Tb terms, then there is another FQ log entry ... but it shows four logs grouped as one. This could only happen if four mail items were being processed by the same search term at the same time. That means processing is Asynchronous, where multiple email items are being processed at the same time, and can be processed by the same filter and the same term at the same time, resulting in a grouping of identical (timestamped) log entries.

I am not sure about those log grouping rules, grouping in console is somewhat annoying because it deserialises debug output which ain’t helpful. When console groups equivalent messages (not sure how it determines equality as it cannot look ahead) I believe it might delay certain types of output to collate later on, so i would take those grouped log entries with a pinch of salt. It just means the messages were bundled, nothing about that the actions it reports where grouped together or finished at the same tine.

**Comment by RealRaven2000** [comment]
> It confirms that a single message gets run through all terms before another message is processed

Only for a single filter; i imagine incoming messages may be pooled together and then ran through all filters… if a filter matches a specific msg and includes the action „stop processing“  that msg will be removed from the pool of
 Messages to be processed further ( by the remaining filters)  - this is one possible way to implement this. The other one would be to go by messages in order of arrival and run all filters one the first msg, then move on to the next etc.

as Add-on that provides special filter actions we cannot assume that it is either way or that actions on an earlier received message have been executed before an email that was received a bit later. Not sure if Tb can process the mails in exact order when there is synching of content with the mail server going on.

however it is save to assume that (from the perspective of a single incoming mail) the filters are started in the visual order they are displayed in on the message filters window.

 > And yet for actions, it processes a single action on multiple messages, then the next action on all messages, etc.

**Comment by RealRaven2000** [comment]
Not sure how i clicked close discussion, when i edit a comment on iAd I sometimes accidentally send off my comment prematurely… maybe i get to trigger close with comment accidentally. Not sure…
